### DISCUSSION TITLE
Question about using in rust

### DISCUSSION THREAD

**Discussion by cheunglok97** [gorules/zen#112] [Q&A] [answered]
```rust
#[tokio::main]
async fn main() {
    let handle = tokio::spawn(async {
        evaluate().await;
    });
    let handle2 = tokio::spawn(async {
        evaluate().await;
    });
    let (result, result2) = tokio::join!(handle, handle2);
}

async fn evaluate() {
    let decision_content: DecisionContent = serde_json::from_str(include_str!("jdm_graph.json")).unwrap();
    let engine = DecisionEngine::default();
    let decision = engine.create_decision(decision_content.into());

    let result = decision.evaluate(&json!({ "input": 12 })).await;
}

``` 

I am trying to studying this project and I and faced some problem when I try it.

Then I compile the above code, I got 
`Rc<UnsafeCell<Vec<&zen_expression::opcodes::Opcode<'_>>>>` cannot be sent between threads safely`

May I know how to sove this? 

Great thanks!

**Comment by stefan-gorules** [comment, marked as answer]
Hi @cheunglok97,

ZEN expressions are optimized for single-threaded use and don't support multi-threading in the evaluation of a single expression. This limitation applies to the ZEN engine as well.

To integrate ZEN decisions within a Tokio-based, multi-threaded environment, you can use `Handle::current().block_on `to execute these decisions synchronously within an asynchronous context. This approach bypasses Tokio's thread safety requirements for asynchronous functions, allowing you to include non-thread-safe ZEN operations in your Tokio application. For a practical implementation, see this [GitHub example](https://github.com/gorules/editor/blob/d55106014e8f2749513a06aaff4fecce7811ef1c/backend/src/main.rs#L79:L97).

Let me know if this solves your issue.

**Reply by cheunglok97** [reply]
Thanks! That is a very good example

**Reply by andymk-dev** [reply]
This worked earlier but it doesn't seem to work with the latest version (0.33). I'm getting the following error again.

`std::rc::Rc<str>` cannot be sent between threads safely
within `std::result::Result<zen_engine::DecisionGraphResponse, std::boxed::Box<zen_engine::EvaluationError>>`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<str>`, which is required by `std::result::Result<zen_engine::DecisionGraphResponse, std::boxed::Box<zen_engine::EvaluationError>>: std::marker::Send`rustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20[2]?2#file:///home/andy/io/pocpc-api/src/routes/api/uw.rs)
mod.rs(21, 10): required because it appears within the type `zen_engine::Variable`
graph.rs(441, 12): required because it appears within the type `zen_engine::DecisionGraphResponse`
result.rs(527, 10): required because it appears within the type `std::result::Result<zen_engine::DecisionGraphResponse, std::boxed::Box<zen_engine::EvaluationError>>`
blocking.rs(205, 12): required by a bound in `tokio::task::spawn_blocking`
